### PR TITLE
Feat/registration

### LINKED ISSUE #10: feat/Fast soelle to member prosess
Should we have a system to make it easy to change status to member right away at Phaesum to avoid a long process of editing each profile later on.

### PR DESCRIPTION
**Pull request by theodorklauritzen**
- Add more registration fields (alergy, phonenumber)
- Add forgot password
- Add verify email
- Add a redirect to /register if the user is not registered, and foes to a page with getUser shouldRedirect is set to true
- Add OmegaId (important milestone for #10)
- Added a way to register soelles to members (Fixes #10)
- Added Qr code libraries ( #139 #41)
- Add UI to admin studyProgrammes
- Automaticly add users to OmegaMembershipsGroups

Caution! Merge the feat/notifications #240 branch and the feat/auth folder #228 first!

### CODE REVIEW

File: src/server/groups/omegaMembershipGroups/read.ts
Comment (Paulijuz) [inline src/server/groups/omegaMembershipGroups/read.ts:50]: wouldn't a more efficient approach be to get the ids of the omega membership groups

`const groupIds = prisma.omegaMembershipGroup.findMany({ select: { id: true } })`

and then check if there exists a membership between the user and any of the groups?

`prisma.memberships.findUnique({ where: { userId, groupId: { OR: groupIds } } })`

**Review by Paulijuz — CHANGES_REQUESTED** [review]
This PR rickrolled me, therefore I cannot approve it. >:(

Also, I added some comments throughout the changes and is some general feedback:
- Are you supposed to be able to enter any domain name you want into a mail alias. Would it be better to simply enter the name that should be before the @?
- Should the admissions be a database table? This doesn't seem like something that is likely to change in a loooong time so I don't see the benefit of allowing it to be dynamic. Maybe a simple enum would suffice? Then you could reduce the complexity quite a bit.
- I'm not sure if this is the PR that did it, but the create new user form is broken.
- A lot of the forms revert to the old values after submission, but I'm guessing a lot of this will be fixed when we do a second pass over the UI.
- I think turning on a lot of notification channels is very tedious. Especially since you have to press the submit button for each channel.

I don't know if I was too picky (a lot of it was just grammar) or not so fix the issues you deem worthwhile to fix and then I'll approve this PR. All inn all this PR brings a lot of critical features. It's really nice to se some good progress.

**Comment by theodorklauritzen** [comment]
Thanks for the feedback @Paulijuz 

> Are you supposed to be able to enter any domain name you want into a mail alias. Would it be better to simply enter the name that should be before the @?

Some users should be able to add different domains than the current domain. Maybe we want the alias POSTMASTER@sanctus.omega.ntnu.no for example. But I agree that it's unnecessary in 99% of cases not to add this automatically and prevent users from typing the wrong domain.

> Should the admissions be a database table? This doesn't seem like something that is likely to change in a loooong time so I don't see the benefit of allowing it to be dynamic. Maybe a simple enum would suffice? Then you could reduce the complexity quite a bit.

I had an idea to use this more generally, like admissions to DGR. But after some thought I don't think this will be used, and then it's just extra unnecessary complexity.

> I'm not sure if this is the PR that did it, but the create new user form is broken.

I have some thoughts about this. I think the user form should take in the fields `username`, `firstname`, `lastname` and `email`. Then when you create a user, they should get an email with a link that leads to the registrations page. And there they can create a password add a mobile number etc. Do you have any thoughts on this as well @JohanHjelsethStorstad ?

> I think turning on a lot of notification channels is very tedious. Especially since you have to press the submit button for each channel.

I totally agree the UI should be a lot better. I was not sure how the profile page design would be. But the best is probably to put the settings on a different page.

**Comment by Paulijuz** [comment]
> Some users should be able to add different domains than the current domain. Maybe we want the alias [POSTMASTER@sanctus.omega.ntnu.no](mailto:POSTMASTER@sanctus.omega.ntnu.no) for example. But I agree that 99% of cases we just use omega.ntnu.no.

Then I understand your design choice. Ideally, I still think there should be something in the UI to indicate which domains are valid tho. Currently you either have to know which domains are valid or look through the list of used domains. Maybe you could have a drop down of all the valid domains? Though this is not really important, this page will be used by very few people (only admins) so I think how it is now is good enough. We can always improve later on.

**Comment by JohanHjelsethStorstad** [comment]
I agree that we should have a fixed number of domains allowed

**Comment by JohanHjelsethStorstad** [comment]
As for the createNewUser form it was meant as an admin-form. So it should probably  take as little info as possible then make the new user give the rest

File: src/server/mail/read.ts
Comment (JohanHjelsethStorstad) [inline src/server/mail/read.ts:78]: This should have some includers. Often (look at cms modules) when we had long includes I started aking them in ConfigVars.

File: src/auth/jwt.ts
Comment (JohanHjelsethStorstad) [inline src/auth/jwt.ts:10]: As I have said I think I will make a directory for jwt

File: src/app/components/User/CreateUserForm.tsx
Comment (JohanHjelsethStorstad) [inline src/app/components/User/CreateUserForm.tsx:29]: med en link

File: src/app/components/OmegaId/identification/OmegaId.tsx
Comment (JohanHjelsethStorstad) [inline src/app/components/OmegaId/identification/OmegaId.tsx:3]: I think all of this needs a generalization to use qr-codes for skapres, registrations, ...

Comment (JohanHjelsethStorstad) [reply src/app/components/OmegaId/identification/OmegaId.tsx:3]: Just remember that it is written I guess

File: src/actions/news/update.ts
Comment (JohanHjelsethStorstad) [inline src/actions/news/update.ts:30]: A little unsure if this should be done in the action or in the function??

File: src/actions/omegaquotes/create.ts
Comment (JohanHjelsethStorstad) [inline src/actions/omegaquotes/create.ts:27]: Same as for news. We just need a convention.

**Review by JohanHjelsethStorstad — COMMENTED** [review]
Looks good

**Review by JohanHjelsethStorstad — CHANGES_REQUESTED** [review]
The above

File: src/server/omegaid/generate.ts
Comment (theodorklauritzen) [reply src/server/omegaid/generate.ts:25]: GenerateJWT uses symmetric keys, this function uses asymmetirc keys. The function should be moved to make this clear

Comment (JohanHjelsethStorstad) [reply src/server/omegaid/generate.ts:25]: Then I think you should have two methods in the jwt file

File: src/app/components/OmegaId/identification/OmegaId.tsx
Comment (theodorklauritzen) [reply src/app/components/OmegaId/identification/OmegaId.tsx:3]: The next-qrcode library has a simple api. I don't think we need to make an abstraction on top of that. See `OmegaIdElement.tsx` for an example.

File: src/server/mail/read.ts
Comment (theodorklauritzen) [reply src/server/mail/read.ts:78]: I can move the include statements to a ConfigVars file, the the include statements will probably only be used one place.

File: src/server/auth/resetPassword.ts
Comment (theodorklauritzen) [reply src/server/auth/resetPassword.ts:8]: Hmm, I have combined the functions. I think it is nice to have the functions sending email in the same folder, but I'm unsure why, so we can change that at place the function from where they were called from.

File: src/styles/_mixins.scss
Comment (theodorklauritzen) [reply src/styles/_mixins.scss:115]: I agree, and it would be nice to discuss how we should implement different screen sizes. They should maybe be called sm, lg instead of mobile as I have written.

**Comment by theodorklauritzen** [comment]
I have fixed most of the feedback you gave. I have also created a JWT folder, with all the JWT functionality.  The things I haven't fixed yet are:

- mobile mixin on all pages
- Includer for long queries (I don't feel like the queries should be in config since they are not config)
- The user form is still broken, but it would be nice to discuss how this should function. (And I is fixed in a later PR)

File: src/server/notifications/email/mailHandler.ts
Comment (Paulijuz) [inline src/server/notifications/email/mailHandler.ts:136]: this should be moved to src/typings/global.d.ts

File: src/server/notifications/email/systemMail/userInvitivation.tsx
Comment (Paulijuz) [inline src/server/notifications/email/systemMail/userInvitivation.tsx:15]: It feels a bit iffy using an environment variable created for next auth elswhere in our project. Especially since you've used `DOMAIN` elsewhere, for example in `dispatch.ts`. Logically we should be using `DOMAIN` and set it to https://localhost:80. Then we could set `NEXTAUTH_URL=DOMAIN` to simplify things. However, I guess it might look a bit weird having localhost plastered all over the website instead of omega.ntnu.no, but that might not be a bad thing because then we could easily spot hard coded URLs. Another option would be to add another environment variable so we could differentiate between the actual link url and the domain which should appear for visual thing only.

File: src/server/users/update.ts
Comment (Paulijuz) [inline src/server/users/update.ts:68]: The condition in this if statement will always be true due to the previous if statement. So it's unnecessary.

File: src/server/notifications/email/mailHandler.ts
Comment (Paulijuz) [reply src/server/notifications/email/mailHandler.ts:136]: you can see how it's done with prisma :)

**Review by Paulijuz — DISMISSED** [review]
Nice work! Really like the new notification settings menu and the comment about how mail flow works is very helpful. I also like the new login color scheme, although we will need a pass over the UI to make all the pages consistent with each other. 

I haven't done as much extensive testing since I'm guessing the core functionality works the same as before. I still managed to find some UI bug tho, all of them are _minor_ things related to the new notification settings page.
- If you make further changes by toggling channels after having pressed save, the save button will still disappear. 
- If I enable all the channels with the top check button I _sometimes_ get the error message "The methods must a subset of the available methods". I'm guessing this is because the top check button checks the disabled check buttons somehow.

I also made some small comment. Keep up the good work!

**Comment by theodorklauritzen** [comment]
The latest comments from the review from @Paulijuz  should have been resolved by now. Hovever I have not bee enable to reproduce the bug:

> If I enable all the channels with the top check button I sometimes get the error message "The methods must a subset of the available methods". I'm guessing this is because the top check button checks the disabled check buttons somehow.

Therefore it's probably not fixed, but I have changed how the page works and that can be why I'm unable to reproduce the bug.

@JohanHjelsethStorstad do you want to give me another review?

File: .eslintignore
Comment (johanMedistim) [inline .eslintignore:6]: O if you look in global.d.ts I used disable no var for one line. Either remove this or ignore the file. Now the ignore is pointless

Comment (johanMedistim) [inline .eslintignore:6]: f

File: src/server/notifications/channel/validation.ts
Comment (johanMedistim) [inline src/server/notifications/channel/validation.ts:53]: This seams a little dangerous

Comment (JohanHjelsethStorstad) [inline src/server/notifications/channel/validation.ts:53]: This feels dangerous

File: src/typings/global.d.ts
Comment (JohanHjelsethStorstad) [inline src/typings/global.d.ts:8]: Above I have ignored the no-var rule. Either add it to this line too and remove the file from .eslintignore, or remove my no-var ignore, since now it is pointless.

File: src/server/notifications/Types.ts
Comment (JohanHjelsethStorstad) [inline src/server/notifications/Types.ts:26]: These are not types?

**Review by JohanHjelsethStorstad — APPROVED** [review]
I think it mostly looks fine now

**Review by JohanHjelsethStorstad — DISMISSED** [review]
It looks fine, but there is logic here that is hard to understand. Also we can fix some small things I have commented

**Comment by theodorklauritzen** [comment]
@JohanHjelsethStorstad fixed your comments.

I'll refactor a few things in this PR while working on push notifications.

**Review by JohanHjelsethStorstad — APPROVED** [review]

**Pull request merged**
